### PR TITLE
Refactor python documentation and update examples

### PR DESCRIPTION
**Pull request by anthony-santana**
### Description

Revives the closed PR #238 for:
* Refactoring all of the python documentation
* Adding noise modeling tests and examples
* Modifies the pybind header to fix bugs in our documentation types

Issues this PR introduces or doesn't solve:
* Free functions and certain members (edit: it appears this applies to everything except classes and properties/attributes) are given two code signatures in the built docs. For example, see the x/y/z/etc. gates in `py_kernel_builder.cpp`. They're given a `x()` at the top, then new-line and indented below is the full signature `x(self: cudaq.Kernel, target: cudaq.QuakeValue)→ None`. I believe this is coming from the `python_api.rst` docs, e.g., when we write `.. automethod:: x`, I believe the "x" there is what's being printed on the top-line of the function doc. This is minor enough to merge right now, and an issue will be filed and resolved in a follow-up PR.

### CODE REVIEW

File: docs/sphinx/examples/cpp/basics/noise_amplitude_damping.cpp [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+// we will examine the modeling of energy dissipation within our system
+// via environmental interactions. The result of this "amplitude damping"
+// is to return the qubit to the |0> state with a user-specified probability.
+
+int main() {
+
+  // We will begin by defining an empty noise model that we will add
+  // our damping channel to.
+  cudaq::noise_model noise;
+
```
Comment (bmhowe23) [inline docs/sphinx/examples/cpp/basics/noise_amplitude_damping.cpp:22]: What do you think about adding `cudaq::set_random_seed(13)` to make this test repeatable? The same question applies to all the new cpp and Python tests in this review. (Similar to the spirit of PR #617)

Comment (bmhowe23) [reply docs/sphinx/examples/cpp/basics/noise_amplitude_damping.cpp:22]: Thank you, and yes, that sounds reasonable to me. While we run the docs examples as part of our CI, I'm ok with a targeted subset of them being non-repeatable. As we make more progress on #608, we may want to consider updating the docs to describe to our users how to make some of the docs tests more repeatable, too, but that doesn't have to be part of this PR. Thanks!

**Review by bettinaheim — APPROVED** [review]
I did a quick read through. All looks good to me.

**Pull request merged**